### PR TITLE
[CINFRA-143] Added establish-leadership-semantics.

### PR DESCRIPTION
**Pull request by maierlars**
### Scope & Purpose
The leadership is established, when the first entry of that term is committed. This is just an intermediate step. Future PRs will integrate this into move shard operations.

We introduce a boolean `leadershipEstablished` that is exposed through the log status and is indirectly reported to the agency by updating Current/ReplicatedLog/Leader/Term to the current term.

### Testing & Verification

*(Please pick either of the following options)*

- [ ] This change is a trivial rework / code cleanup without any test coverage.
- [x] The behavior in this PR was *manually tested*
- [x] This change is already covered by existing tests, such as all existing Replication2 tests.
- [x] This PR adds tests that were used to verify all changes:
  - x] Added new C++ **Unit tests**
  - [ ] Added new **integration tests** (e.g. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)

### CODE REVIEW

**Review by mpoeter — APPROVED** [review]
Some minor comments, otherwise LGTM!

File: arangod/Replication2/ReplicatedLog/LogLeader.cpp [hunk 1]
```diff
@@ -1030,6 +1030,44 @@ auto replicated_log::LogLeader::LocalFollower::resign() && noexcept
...
+        LOG_CTX("536f4", INFO, self->_logContext) << "leadership established";
+      } catch (std::exception const& err) {
+        LOG_CTX("5ceda", FATAL, self->_logContext)
+            << "failed to establish leadership: " << err.what();
+      }
+    }
+
+    LOG_TOPIC("94696", TRACE, Logger::REPLICATION2)
+        << "leader is already gone, no leadership was established";
+  });
```
[the remark this answers is not in the document]
Comment (maierlars) [reply arangod/Replication2/ReplicatedLog/LogLeader.cpp:1068]: That's true.

File: arangod/Cluster/Maintenance.cpp [hunk 2]
```diff
@@ -1557,42 +1573,83 @@ arangodb::Result arangodb::maintenance::reportInCurrent(
...
+              VPackObjectBuilder o(&report);
+              report.add(OP, VP_SET);
+              report.add("payload", VPackValue(localTerm.value().value));
+              {
+                VPackObjectBuilder preconditionBuilder(&report, "precondition");
+                report.add(preconditionPath, VPackValue(localTerm.value().value));
+              }
             }
           }
         }
```
[the remark this answers is not in the document]
Comment (maierlars) [reply arangod/Cluster/Maintenance.cpp]: Done

File: arangod/Replication2/ReplicatedLog/LogLeader.cpp [hunk 3]
```diff
@@ -1030,6 +1030,43 @@ auto replicated_log::LogLeader::LocalFollower::resign() && noexcept
...
+    auto firstIndex = data.insertInternal(std::nullopt, true, std::nullopt);
+    TRI_ASSERT(firstIndex == lastIndex.index + 1);
+    return firstIndex;
+  });
+
+  waitFor(waitForIndex).thenFinal([weak = weak_from_this()](futures::Try<WaitForResult>&& result) noexcept {
+    if (auto self = weak.lock(); self) {
+      try {
+        result.throwIfFailed();
+        self->_guardedLeaderData.getLockedGuard()->_leadershipEstablished = true;
```
Comment (apetenchea) [inline arangod/Replication2/ReplicatedLog/LogLeader.cpp:1057]: I guess the opposite of this should happen in `LogLeader::resign`. Not sure if you intended to do that in another PR.

**Review by apetenchea — COMMENTED** [review]
Looks good. I left some small remarks.

File: arangod/Cluster/Maintenance.cpp [hunk 4]
```diff
@@ -1198,6 +1198,135 @@ static VPackBuilder assembleLocalDatabaseInfo(DatabaseFeature& df, std::string c
...
+  LOG_CTX("11dbd", TRACE, logContext)
+      << "checking replicated log " << id
+      << " local term = " << (localTerm ? to_string(*localTerm) : "n/a") << " current "
+      << (!termInCurrent.isNone() ? termInCurrent.toJson() : "n/a") << " Current/Leader/Term "
+      << (!leaderTerm.isNone() ? leaderTerm.toJson() : "n/a");
+
+  if (requireUpdateLocal) {
+    // Check Current/ReplicatedLogs/<dbname>/<logId>/localStatus/<serverId>/currentTerm != currentTerm
+    // if so, update Current/ReplicatedLogs/<dbname>/<logId>/localStatus/<serverId> with
+    //  {"currentTerm": currentTerm, "spearHead": {"index": last-index, "term": last-term}}
```
[the remark this answers is not in the document]
Comment (maierlars) [reply arangod/Cluster/Maintenance.cpp:1257]: Yes

File: arangod/Replication2/ReplicatedLog/LogLeader.cpp (hunk 3 above)
Comment (maierlars) [reply arangod/Replication2/ReplicatedLog/LogLeader.cpp:1057]: I am against this, because once the leadership has been established for a given term, it will never become not established. So this is in fact a sticky bit.

File: tests/Replication2/ReplicatedLog/AppendEntriesBatchTest.cpp [hunk 5]
```diff
@@ -81,6 +81,14 @@ TEST_P(AppendEntriesBatchTest, test_with_sized_batches) {
...
       }
     }
+    {
+      // Add first entry in term
+      currentSize += PersistingLogEntry{LogTerm{5}, LogIndex{1}, std::nullopt}.approxByteSize();
+      if (currentSize >= _optionsMock->_maxNetworkBatchSize) {
+        numRequests += 1;
+        currentSize = 0;
+      }
+    }
```
[the remark this answers is not in the document]
Comment (maierlars) [reply tests/Replication2/ReplicatedLog/AppendEntriesBatchTest.cpp:91]: Ha! No! You got me in the first place, but this is not the same. We add a new message whenever we go beyond the threshold but we do not account for how much. For your formula to hold we would have to split up payloads in different messages.
